examples/runner-custom-terraform-spot-fleet/main.tf

```diff
@@ -0,0 +1,265 @@
+    # Don't forget about trailing slash, which indicates that configuration is a folder and not a single file
+    "terraform-config=/extra-files/runners-config.zip/",
+    "terraform-variables-from=/extra-files/runners-vars.json",
+    "engine-storage-driver=overlay2"
+  ]
+
+  # First, lets include a bunch of files:
+  # - terraform configuration which is used by terraform docker-machine driver to deploy runner instance
+  # - script which is used to download and install terraform and docker-machine-driver-terraform
+  # - configuration file for terraform-created runners
+  extra_files = {
+    "runners-config/main.tf"      = file("${path.module}/worker-config/main.tf")
+    "runners-config/variables.tf" = file("${path.module}/worker-config/variables.tf")
+    "runners-config/outputs.tf"   = file("${path.module}/worker-config/outputs.tf")
+    "docker-machine-terraform.sh" = templatefile("${path.module}/docker-machine-terraform.sh", {
+      terraform_url = "https://releases.hashicorp.com/terraform/0.13.3/terraform_0.13.3_linux_amd64.zip"
+      # This is a fork of @tintoy's docker-machine driver, updated to make TF >=0.12 & <0.14 run with it
+      terraform_driver_url = "https://github.com/krzysztof-miemiec/docker-machine-driver-terraform/releases/download/v0.4.2/linux-amd64.zip"
+    })
+
+    # This corresponds to variables.tf in worker-config folder
+    # Keep in mind that variables set below do not affect instances created using Terraform
+    # and everything has to be set (again) here
+    "runners-vars.json" = jsonencode({
+      image_id                    = data.aws_ami.ubuntu.image_id
+      instance_types              = local.instance_types
+      region                      = var.aws_region
+      subnets                     = local.subnet_ids
+      security_group_ids          = [module.gitlab_runner.runner_sg_id]
+      spot_price                  = "0.2"
+      volume_size                 = 50
+      iam_instance_profile        = aws_iam_instance_profile.runner.name
+      spot_fleet_tagging_role_arn = aws_iam_role.spot_fleet_tagging.arn
+      tags = {
+        "tf-aws-gitlab-runner:instancelifecycle" = "spot:yes"
+      }
+    })
+  }
+
+  # Let's run Terraform & docker-machine-driver-terraform installation script on boot
+  userdata_pre_install = <<SCRIPT
+    chmod +x /extra-files/docker-machine-terraform.sh
+    /extra-files/docker-machine-terraform.sh
+    # We have to pass TF config as zip file, because docker-machine-driver-terraform symlinks a folder if passed directly
+    (cd /extra-files/runners-config; zip -r ../runners-config.zip *)
+  SCRIPT
+
+  # After each new configuration change that happens on S3, don't forget to reload runners-config
+  # extra_files we passed earlier, are pulled to `/extra-files` directory on Gitlab Agent instance
+  # Let's zip the contents of runner config after each config update
+  post_reload_config = "(cd /extra-files/runners-config; zip -r ../runners-config.zip *)"
```

Comment:
The command here probably should be `zip -r ../runners-config.zip .`. If we use the wildcard `*`, this will not pick up any hidden files. An example of a useful hidden file we might have is `.terraform.lock.hcl` file which is used from terraform 0.13. Specifying it allows minimise surprises so it's quite nice to include.